### PR TITLE
#8264: Worker thread optimizations:

### PR DESCRIPTION
**Pull request by tt-asaigal**
- Remove cross worker syncs, fastest worker now populates metadata
  - Use atomic flags to track if a tensor is populated instead of using a vector of booleans
  - Worker thread binding is now NUMA aware
  - Main thread + python/torch threads are also bound to cores not assigned to workers (more deterministic perf)
  - Improve multi-device synchronize to remove main thread overhead

**Lots of clang formatting makes this PR massinve**

### CODE REVIEW

**Comment by tt-asaigal** [comment]
Passing MultiChip: https://github.com/tenstorrent/tt-metal/actions/runs/9214999624/job/25352548694
Passing Post commit: https://github.com/tenstorrent/tt-metal/actions/runs/9214984431

**Comment by TT-billteng** [comment]
do you actually need to bump UMD for this PR?

**Comment by tt-asaigal** [comment]
> do you actually need to bump UMD for this PR?

Yeah this commit needs a bump. We're relying on a new API that gives us a PCIe device to NUMA node mapping.

**Review by TT-billteng — APPROVED** [review]

**Review by xanderchin — APPROVED** [review]

File: tt_eager/tensor/types.hpp [hunk 1]
```diff
@@ -449,7 +449,8 @@ struct MultiDeviceHostStorage {
         std::vector<int> ordered_device_ids;
         std::unordered_map<int, DeviceBuffer> buffers;
         std::unordered_map<int, Shape> shapes;
-        mutable std::mutex mtx;
+        mutable std::mutex buffer_mtx;
```
Comment (cfjchu) [inline tt_eager/tensor/types.hpp:452]: Curious how much did this change improved e2e perf?

File: tt_eager/tensor/tensor.hpp [hunk 2]
```diff
@@ -32,10 +32,12 @@ struct Tensor {
         DataType dtype;
         Layout layout;
         std::mutex populated_mutex;
-        std::vector<bool> tensor_populated = {};
+        uint32_t num_shards = 0;
```
Comment (cfjchu) [inline tt_eager/tensor/tensor.hpp:35]: can we use a better variable name here

File: tt_eager/tensor/tensor_impl.hpp [hunk 3]
```diff
@@ -385,7 +385,6 @@ inline Tensor to_host(const Tensor& tensor, bool blocking = true) {
             host_tensor.set_dtype(tensor.get_dtype());
             host_tensor.set_layout(tensor.get_layout());
             insert_buffer_and_shape_for_device(device, shard, host_tensor, device_index);
-            host_tensor.set_populated(device);
```
Comment (cfjchu) [inline tt_eager/tensor/tensor_impl.hpp:388]: why is this no longer needed?

Comment (tt-asaigal) [reply tt_eager/tensor/tensor_impl.hpp:388]: This function no longer exists. Instead, we check the number of workers that have completed and compare them against the number of shards in the tensor. `tensor.cpu()` which eventually calls this function does that handling.

File: tt_eager/tensor/tensor.hpp (hunk 2 above)
Comment (tt-asaigal) [reply tt_eager/tensor/tensor.hpp:35]: I will address this in the next commit. Does `num_shards_to_be_populated` seem better? That's effectively what we're tracking here.

File: tt_eager/tensor/types.hpp (hunk 1 above)
Comment (tt-asaigal) [reply tt_eager/tensor/types.hpp:452]: ~2-3% ... which seems like noise but it seemed consistently better across runs.

File: module.mk [hunk 4]
```diff
@@ -89,7 +89,8 @@ LDFLAGS ?= $(CONFIG_LDFLAGS) -L$(LIBDIR) \
 	-lboost_regex \
 	-lpthread \
 	-latomic \
-	-lhwloc
+	-lhwloc \
+	-lnuma
```
[the remark this answers is not in the document]
Comment (vtangTT) [reply module.mk:93]: package is already there when i tried locally

File: CMakeLists.txt [hunk 5]
```diff
@@ -14,6 +14,10 @@ project(tt-metal
 find_package(Boost REQUIRED COMPONENTS thread filesystem system regex)
 find_package(GTest REQUIRED)
 find_package (Python3 COMPONENTS Interpreter Development)
+find_library(NUMA_LIBRARY NAMES numa)
+if (NOT NUMA_LIBRARY)
```
Comment (TT-billteng) [inline CMakeLists.txt:18]: does the `REQUIRED` parameter not work here?

File: tt_metal/CMakeLists.txt [hunk 6]
```diff
@@ -18,7 +18,7 @@ set(TT_METAL_OBJECTS
 
 add_library(tt_metal ${TT_METAL_OBJECTS})
 if(BUILD_SHARED_LIBS)
-    target_link_libraries(tt_metal PUBLIC device)
+    target_link_libraries(tt_metal PUBLIC device metal_common_libs)
```
Comment (tt-rkim) [inline tt_metal/CMakeLists.txt:21]: @vtangTT This is crazy - we didn't need these libs for metal? Wow - guess everything is taken care of by umd

File: CMakeLists.txt (hunk 5 above)
Comment (tt-rkim) [reply CMakeLists.txt:18]: I would also prefer `REQUIRED`

**Review by cfjchu — APPROVED** [review]

File: tt_metal/tt_metal.cpp [hunk 7]
```diff
@@ -171,6 +172,78 @@ std::vector<Device *> devices;
...
+        }
+    } else {
+        // System does not use NUMA. Use-round robin binding strategy.
+        core_assigned_to_device = mmio_controlled_device_id % sysconf(_SC_NPROCESSORS_ONLN);
+    }
+    free_cores.erase(core_assigned_to_device);
+    return core_assigned_to_device;
+}
+
+void bind_current_thread_to_free_cores(const std::unordered_set<uint32_t> &free_cores) {
```
Comment (pgkeller) [inline tt_metal/tt_metal.cpp:228]: is this code called at init or during runtime?  I find it ironic that CPU_SET (etc) use a bitmask for cpus but we use an unordered set, these are orders of magnitude apart in terms of efficiency.  In general, I dislike how much we rely on sets and maps when bitmasks or arrays will work dramatically faster w/ dramatically less cache footprint.  I haven't wrapped my head around this enough to know if this is another instance but will raise the question here.

Comment (tt-asaigal) [reply tt_metal/tt_metal.cpp:228]: This is during init only. We set thread affinities once at the start. I get your point though, we can use bit-masks here ... just makes some of the code look a bit ugly and wasn't initially intuitive so I used maps instead.

**Review by pgkeller — APPROVED** [review]

**Pull request merged**